src/main/host/descriptor/mod.rs

```diff
@@ -1072,6 +1075,49 @@ mod export {
 
         file.canonical_handle()
     }
+
+    #[no_mangle]
+    pub extern "C" fn eventsource_new() -> *mut RootedRefCell_StateEventSource {
+        let event_source = worker::Worker::with_active_host(|host| {
+            Box::new(RootedRefCell::new(host.root(), StateEventSource::new()))
+        })
+        .unwrap();
+        Box::into_raw(event_source)
+    }
+
+    #[no_mangle]
+    pub extern "C" fn eventsource_free(event_source: *mut RootedRefCell_StateEventSource) {
+        assert!(!event_source.is_null());
+        unsafe { Box::from_raw(event_source) };
+    }
+
+    #[no_mangle]
+    pub extern "C" fn eventsource_addLegacyListener(
+        event_source: *mut RootedRefCell_StateEventSource,
+        listener: *mut c::StatusListener,
+    ) {
+        let event_source = unsafe { event_source.as_mut() }.unwrap();
```

Comment:
IMO this should take a `*const RootedRefCell_StateEventSource`, and here we should use `as_ref` instead of `as_mut`. This helps ensure we don't accidentally create incompatible references.

Reply:
Nice catch, fixed.